Source/playerdat.cpp

```diff
@@ -7,68 +7,272 @@
+	{
+	}
+
+	bool operator==(const ExperienceColumnDefinition &other) const
+	{
+		return type == other.type && skipLength == other.skipLength;
+	}
 };
 
+void ReloadExperienceData()
+{
+	LoadDataResult dataFile = LoadDataFile("txtdata\\Experience.txt");
+	if (dataFile.hasError()) {
+		// Can't load the data file, so reset to defaults.
+		// TODO: Log error somewhere the player can see it, they probably want to know that we couldn't load the file
+		ExperienceData = {};
+		return;
+	}
+
+	constexpr unsigned ExpectedColumnCount = 2; // Level, Experience
+
+	std::vector<ExperienceColumnDefinition> columns;
+	columns.reserve(ExpectedColumnCount);
+	std::unordered_set<ExperienceColumnDefinition::ColumnType> seenColumns;
+
+	unsigned currentColumn = 0;
+	unsigned lastColumn = 0;
+	GetFieldResult result { dataFile.begin() };
+	do {
+		if (columns.size() >= ExpectedColumnCount) {
+			// All key columns have been identified and there's extra fields, discard them and proceed
+			result = DiscardRemainingFields(result.ptr, dataFile.end());
+			break;
+		}
+
+		std::string headerField;
+		result = GetNextField(result.ptr, dataFile.end(), headerField);
+		// As long as we don't call GetNextField twice it a row it will always contain a value
+		ExperienceColumnDefinition::ColumnType columnType = ExperienceColumnDefinition::mapNameToType(headerField);
+		if (columnType != ExperienceColumnDefinition::ColumnType::Unknown && seenColumns.find(columnType) == seenColumns.end()) {
+			unsigned skipColumns = 0;
+			if (currentColumn > lastColumn)
+				skipColumns = currentColumn - lastColumn - 1;
+			columns.emplace_back(columnType, skipColumns);
+			lastColumn = currentColumn;
+		}
+		++currentColumn;
+	} while (result.ok());
+
+	if (result.eof()) {
+		// The data file ended after the header, since there's no data we need to reset to defaults.
+		// TODO: Log error somewhere the player can see it, they probably want to know that their file is missing data
+		ExperienceData = {};
+		return;
+	}
+
+	if (columns.size() < ExpectedColumnCount) {
+		// The data file doesn't have the required headers. Again reset to default (though we could potentially just allocate missing columns in the default order?)
+		// TODO: Log error somewhere the player can see it, they probably want to know that their file has the wrong headers
+		ExperienceData = {};
+		return;
+	}
+
+	ExperienceData.levelThresholds.clear();
+	bool foundMaxLevelRecord = false;
+	do {
+		uint8_t level = 0;
+		uint32_t experience = 0;
+		bool isMaxLevelRecord = false;
+		for (auto &column : columns) {
+			result = DiscardMultipleFields(result.ptr, dataFile.end(), column.skipLength);
+
+			if (!result.ok()) {
+				// reached the end of record early
+				// TODO: let the player know their data file is incomplete
+				break;
+			}
+
+			switch (column.type) {
+			case ExperienceColumnDefinition::ColumnType::Level: {
+				auto fromCharsResult = from_chars(result.ptr, dataFile.end(), level);

```

Comment:
Consider using `ParseInt` with max 255, because the level is represented as a `uint8_t` internally.

Reply:
You already pointed out the reasoning in the later comment but just to keep in in context: the end-of-parsing pointer in the from_chars return value is used in the next call to GetNextField/DiscardField to save scanning values twice. It's a micro-optimisation that would have minimal impact in the real world so I'd be fine with dropping it and just using the helper you wrote, or even updating ParseInt to also provide a new end pointer in the return value.

I'll leave it as-is for now.